[pt2][inductor] update choice caller hashes

nmacchioni opened the pull request:
Summary:
update the hashing method for `ChoiceCaller` class.

`TritonTemplateCaller` objects will now be hashed to:
`{name}-({BLOCK_M}, {BLOCK_N}, {BLOCK_K})-{num_stages}-{num_warps}-{code_hash}`

for example:
`triton_mm-(64, 32, 32)-4-8-cptlntwzcl2gaaofd2oabdwhaqv4ox3lluvbuxitjfhhpz6cyl4o`

`ExternKernelCaller` objects will now be hashed to:
`{name}-{kwargs.keys()[0]}={kwargs.vals()[0]}-...-{code_hash}`

for example:
`addmm-alpha=1-beta=1-c4xxd3iocu4yt6z4udrlqnumays7q6mfnfd3qprh4fxgsvyhqdkf`

Test Plan: sandcastle

Differential Revision: D43285470



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -485,11 +485,7 @@ def _load_library(path):
-            "cpp",
-            extra=cpp_compile_command("i", "o", vec_isa=picked_vec_isa),
-        )
+        key, input_path = write(source_code, "cpp")
```
We need to include the `cpp_compile_command` in the hash so we recompile if it changes.

nmacchioni replied:
included `code_hash(repr(cpp_compile_command(...)))` as part of the hash

jansel commented on torch/_inductor/select_algorithm.py:
```diff
@@ -380,7 +380,10 @@ def generate(
-            mod = PyCodeCache.load(code)
+            mod = PyCodeCache.load(
+                code,
+                extra=f"{kwargs['BLOCK_M'], kwargs['BLOCK_K'], kwargs['BLOCK_K']}-{num_stages}-{num_warps}-",
```
Should not assume the kwargs only contains BLOCK_M/BLOCK_K.  

BLOCK_K is listed twice.

nmacchioni replied:
Oh wow I missed that, one is meant to be BLOCK_N. Let me fix that real quick.

bertmaher reviewed (approved):
lgtm and seems like @jansel's comments were addressed, so 🚢

jansel commented on torch/_inductor/select_algorithm.py:
```diff
@@ -380,7 +380,10 @@ def generate(
-            mod = PyCodeCache.load(code)
+            mod = PyCodeCache.load(
+                code,
+                extra=f"{kwargs['BLOCK_M'], kwargs['BLOCK_N'], kwargs['BLOCK_K']}-{num_stages}-{num_warps}-",
```
What if the config doesn't contain BLOCK_K -- or contains something else like BLOCK_G?

nmacchioni replied:
I've never seen BLOCK_G, but I'll just make it dump all kwargs to be safe (although this will be much more noisy).

jansel replied:
Currently, it doesn't happen, but we want to future proof this code.

nmacchioni committed:
[pt2][inductor] update choice caller hashes (#94853)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/94853

update the hashing method for `ChoiceCaller` class.

`TritonTemplateCaller` objects will now be hashed to:
`{name}-{kwargs.keys()[0]}={kwargs.values()[0]}-...-{num_stages}-{num_warps}-{code_hash}`

for example:
`triton_mm-GROUP_M=8-EVEN_K=False-ALLOW_TF32=True-ACC_TYPE='tl.float32'-BLOCK_M=64-BLOCK_N=64-BLOCK_K=32-num_stages=2-num_warps=4-csdyql3rzthkm6k6jsqt3xndagq6smt2laq7aouvnmembztgpkol`

`ExternKernelCaller` objects will now be hashed to:
`{name}-{kwargs.keys()[0]}={kwargs.values()[0]}-...-{code_hash}`

for example:
`addmm-alpha=1-beta=1-c4xxd3iocu4yt6z4udrlqnumays7q6mfnfd3qprh4fxgsvyhqdkf`

Test Plan: sandcastle

Reviewed By: bertmaher, jiawenliu64

Differential Revision: D43285470

fbshipit-source-id: 2d9518500c48158fe05370277a21f591c879f81d
